lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +678,140 @@ def write(chunk)
+            error_details_map = {}
+            if @partial_success
+              error_details_map = construct_error_details_map(
+                extract_log_entry_errors(error))
+            end
+            if error_details_map.empty?
+              increment_dropped_entries_count(entries_count)
+              @log.warn "Dropping #{entries_count} log message(s)",
+                        error_class: error.class.to_s, error: error.to_s
+            else
+              error_details_map.each do |(error_code, error_message), indexes|
+                partial_error_count = indexes.length
+                increment_dropped_entries_count(partial_error_count)
+                @log.warn "Dropping #{partial_error_count} log message(s)",
+                          error_code: "google.rpc.Code.#{error_code}",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are the indexes of the failed log entries in
+    # the original json payload. The values in this map are the error code and
+    # error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
+    #     "message": "User not authorized.",
+    #     "status": "PERMISSION_DENIED",
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": "[ORIGINAL ERROR] generic::permission_denied: User not a
+    #           uthorized. [google.rpc.error_details_ext] { message: \"User not
+    #           authorized.\" details { type_url: \"type.googleapis.com/google.l
+    #           ogging.v2.WriteLogEntriesPartialErrors\" value: \"\\n\\034\\010\
+    #           \000\\022\\030\\010\\007\\022\\024User not authorized.\\n-\\010\
+    #           \001\\022)\\010\\003\\022%Log name contains illegal character :\
+    #           " } }"
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # In this exxample, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = parse_json_or_nil(error.body)['error']['details']
```

Comment:
I wouldn't use `parse_json_or_nil` -- that function is used for JSON detection, and here we know the body is actually valid JSON. Plus, you lose error information. Just use `JSON.parse` directly, and rescue `JSON::ParserError` instead of a broad catch-all.

Reply:
Changed.